asterius/src/Asterius/CodeGen.hs

```diff
@@ -1516,59 +1538,74 @@ marshalCmmBlockBranch instr = case instr of
-  (br_helper_exprs, maybe_switch_cond_expr, br_branches) <-
+  (br_helper_exprs, maybe_switch_cond_expr, br_branches, needs_unreachable) <-
     marshalCmmBlockBranch exit_node
   pure $ case maybe_switch_cond_expr of
-    Just switch_cond_expr -> RelooperBlock
-      { addBlock = AddBlockWithSwitch
-          { code = concatExpressions $ inner_exprs <> br_helper_exprs,
-            condition = switch_cond_expr
+    Just switch_cond_expr ->
+      ( RelooperBlock
+          { addBlock = AddBlockWithSwitch
+              { code = concatExpressions $ inner_exprs <> br_helper_exprs,
+                condition = switch_cond_expr
+              },
+            addBranches = br_branches
           },
-        addBranches = br_branches
-      }
-    _ -> RelooperBlock
-      { addBlock = AddBlock
-          { code = concatExpressions $ inner_exprs <> br_helper_exprs
+        needs_unreachable
+      )
+    _ ->
+      ( RelooperBlock
+          { addBlock = AddBlock
+              { code = concatExpressions $ inner_exprs <> br_helper_exprs
+              },
+            addBranches = br_branches
           },
-        addBranches = br_branches
-      }
+        needs_unreachable
+      )
   where
     concatExpressions es = case es of
       [] -> Nop
       [e] -> e
       _ -> Block {name = "", bodys = es, blockReturnTypes = []}
 
+unreachableRelooperBlock :: RelooperBlock
+unreachableRelooperBlock = RelooperBlock
+  { addBlock = AddBlock
+      { code = Barf
+          { barfMessage = "unreachable block",
+            barfReturnTypes = []
+          }
+      },
+    addBranches = []
+  }
+
 marshalCmmProc :: GHC.CmmGraph -> CodeGen Function
 marshalCmmProc GHC.CmmGraph {g_graph = GHC.GMany _ body _, ..} = do
   entry_k <- marshalLabel g_entry
-  rbs <-
-    for (GHC.bodyList body) $ \(lbl, GHC.BlockCC _ inner_nodes exit_node) -> do
-      k <- marshalLabel lbl
-      b <- marshalCmmBlock (GHC.blockToList inner_nodes) exit_node
-      pure (k, b)
-  let blocks_unresolved =
-        ( "__asterius_unreachable",
-          RelooperBlock
-            { addBlock = AddBlock
-                { code = Barf
-                    { barfMessage = "unreachable block",
-                      barfReturnTypes = []
-                    }
-                },
-              addBranches = []
-            }
-        )
-          : rbs
+  (needs_unreachable, rbs) <- do
+    let fn ::
+          NeedsUnreachableBlock ->
+          (GHC.Label, GHC.Block GHC.CmmNode GHC.C GHC.C) ->
+          CodeGen (NeedsUnreachableBlock, (BS.ByteString, RelooperBlock))
+        fn needs_unreachable_acc (lbl, GHC.BlockCC _ inner_nodes exit_node) = do
+          k <- marshalLabel lbl
+          (b, needs_unreachable) <- marshalCmmBlock (GHC.blockToList inner_nodes) exit_node
+          pure (needs_unreachable_acc <> needs_unreachable, (k, b))
+    mapAccumLM fn mempty (GHC.bodyList body)
+  let blocks_unresolved
+        | needs_unreachable == coerce True =
```

Comment:
We can do `| coerce needs_unreachable = ...` here. And thus I think there's no need for `Eq` and `Show` instances for `NeedsUnreachableBlock`